### ISSUE TITLE
Gradle Help Task to display configurations

### ISSUE THREAD

**Issue by nniesen** [gradle/gradle#4687]
Add a built-in 'configurations' task to the 'Help' group that displays the configurations and artifacts for the project. It would be a generally useful built-in task for understanding build configurations, their relationships, and their artifacts.

### Custom task that produces the "desired" information:
```
project.tasks.create("configurations") { Task task ->
    task.group = "Help"
    task.description = "Displays the configurations and artifacts of " + project
    task.doFirst {
        println "------------------------------------------------------------"
        println "${project.toString().capitalize()} - ${project.description}"
        println "------------------------------------------------------------"
        configurations.findAll().each { config ->
            println ""
            println "${config.name} - ${config.description}"
            if (!config.extendsFrom.empty) {
                println "    extends from: " + config.extendsFrom.collect({c -> c.name}).join(", ")
            }
            if (!config.allArtifacts.empty) {
                println "    artifacts:"
                config.allArtifacts.files.each { file ->
                    def inherited = (!config.artifacts.files.contains(file)) ? "(*)" : "   "
                    println "         ${inherited} ${project.relativePath(file)}"
                }
            }
        }
        println ""
        println "(*) - inherited artifact"
    }
}
```

### Sample output from a project named buildship created using `init --type java-library`:
```
C:\dev\projects\gradle\buildship>gradlew configurations

> Task :configurations
------------------------------------------------------------
Root project 'buildship' - null
------------------------------------------------------------

annotationProcessor - Annotation processors and their dependencies for source set 'main'.

api - API dependencies for source set 'main'.
    extends from: compile

apiElements - API elements for main.
    extends from: runtime, api
    artifacts:
             build\libs\buildship.jar

archives - Configuration for archive artifacts.
    artifacts:
             build\libs\buildship.jar

compile - Dependencies for source set 'main' (deprecated, use 'implementation ' instead).

compileClasspath - Compile classpath for source set 'main'.
    extends from: compileOnly, implementation

compileOnly - Compile only dependencies for source set 'main'.

default - Configuration for default artifacts.
    extends from: runtimeElements
    artifacts:
         (*) build\libs\buildship.jar

implementation - Implementation only dependencies for source set 'main'.
    extends from: compile, api

runtime - Runtime dependencies for source set 'main' (deprecated, use 'runtimeOnly ' instead).
    extends from: compile
    artifacts:
             build\libs\buildship.jar

runtimeClasspath - Runtime classpath of source set 'main'.
    extends from: runtimeOnly, runtime, implementation
    artifacts:
         (*) build\libs\buildship.jar

runtimeElements - Elements of runtime for main.
    extends from: implementation, runtimeOnly, runtime
    artifacts:
             build\libs\buildship.jar

runtimeOnly - Runtime only dependencies for source set 'main'.

testAnnotationProcessor - Annotation processors and their dependencies for source set 'test'.

testCompile - Dependencies for source set 'test' (deprecated, use 'testImplementation ' instead).
    extends from: compile

testCompileClasspath - Compile classpath for source set 'test'.
    extends from: testCompileOnly, testImplementation

testCompileOnly - Compile only dependencies for source set 'test'.

testImplementation - Implementation only dependencies for source set 'test'.
    extends from: testCompile, implementation

testRuntime - Runtime dependencies for source set 'test' (deprecated, use 'testRuntimeOnly ' instead).
    extends from: testCompile, runtime
    artifacts:
         (*) build\libs\buildship.jar

testRuntimeClasspath - Runtime classpath of source set 'test'.
    extends from: testRuntimeOnly, testRuntime, testImplementation
    artifacts:
         (*) build\libs\buildship.jar

testRuntimeOnly - Runtime only dependencies for source set 'test'.
    extends from: runtimeOnly

(*) - inherited artifact


BUILD SUCCESSFUL in 1s
1 actionable task: 1 executed
```

**Comment by nniesen** [comment]
Sample output for a project named 'core' that depends on a project named 'test' where core produces artifacts: core-2.0.0-SNAPSHOT.jar, core-2.0.0-SNAPSHOT-sources.jar, core-testsupport-2.0.0-SNAPSHOT.jar, and core-testsupport-2.0.0-SNAPSHOT-sources.jar.
```
C:\dev\projects\gradle\core>gradlew configurations
settings.gradle: Local projects included from 'C:\dev\projects\gradle': [test]

> Configure project :
Gradle 4.6 build for project 'core' version '2.0.0-SNAPSHOT' on branch 'develop'.
Found non-jar project dependency: DefaultProjectDependency{dependencyProject='project ':test'', configuration='default'}

> Configure project :test
Gradle 4.6 build for project 'test' version '1.9.0-SNAPSHOT' on branch 'develop'.

> Task :configurations
------------------------------------------------------------
Root project 'core' - Core project library that all projects should include as a compile dependency.
------------------------------------------------------------

annotationProcessor - Annotation processors and their dependencies for source set 'main'.

apiElements - API elements for main.
    extends from: runtime
    artifacts:
             build\dist\core-2.0.0-SNAPSHOT.jar

archives - Configuration for archive artifacts.
    artifacts:
             build\dist\core-2.0.0-SNAPSHOT.jar
             build\dist\core-2.0.0-SNAPSHOT-sources.jar

compile - Dependencies for source set 'main' (deprecated, use 'implementation ' instead).

compileClasspath - Compile classpath for source set 'main'.
    extends from: compileOnly, implementation

compileOnly - Compile only dependencies for source set 'main'.

default - Configuration for default artifacts.
    extends from: runtimeElements
    artifacts:
         (*) build\dist\core-2.0.0-SNAPSHOT.jar

implementation - Implementation only dependencies for source set 'main'.
    extends from: compile

intTestAnnotationProcessor - Annotation processors and their dependencies for source set 'int test'.

intTestCompile - Dependencies for source set 'int test' (deprecated, use 'intTestImplementation ' instead).
    extends from: testSupportCompile
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

intTestCompileClasspath - Compile classpath for source set 'int test'.
    extends from: intTestCompileOnly, intTestImplementation
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

intTestCompileOnly - Compile only dependencies for source set 'int test'.

intTestImplementation - Implementation only dependencies for source set 'int test'.
    extends from: intTestCompile
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

intTestRuntime - Runtime dependencies for source set 'int test' (deprecated, use 'intTestRuntimeOnly ' instead).
    extends from: testSupportRuntime, intTestCompile
    artifacts:
         (*) build\dist\core-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

intTestRuntimeClasspath - Runtime classpath of source set 'int test'.
    extends from: intTestRuntimeOnly, intTestRuntime, intTestImplementation
    artifacts:
         (*) build\dist\core-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

intTestRuntimeOnly - Runtime only dependencies for source set 'int test'.

jacocoAgent - The Jacoco agent to use to get coverage data.

jacocoAnt - The Jacoco ant tasks to use to get execute Gradle tasks.

licenses - null

runtime - Runtime dependencies for source set 'main' (deprecated, use 'runtimeOnly ' instead).
    extends from: compile
    artifacts:
             build\dist\core-2.0.0-SNAPSHOT.jar

runtimeClasspath - Runtime classpath of source set 'main'.
    extends from: runtimeOnly, runtime, implementation
    artifacts:
         (*) build\dist\core-2.0.0-SNAPSHOT.jar

runtimeElements - Elements of runtime for main.
    extends from: implementation, runtimeOnly, runtime
    artifacts:
             build\dist\core-2.0.0-SNAPSHOT.jar

runtimeOnly - Runtime only dependencies for source set 'main'.

testAnnotationProcessor - Annotation processors and their dependencies for source set 'test'.

testCompile - Dependencies for source set 'test' (deprecated, use 'testImplementation ' instead).
    extends from: compile, testSupportCompile
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testCompileClasspath - Compile classpath for source set 'test'.
    extends from: testCompileOnly, testImplementation
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testCompileOnly - Compile only dependencies for source set 'test'.

testImplementation - Implementation only dependencies for source set 'test'.
    extends from: testCompile, implementation
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testRuntime - Runtime dependencies for source set 'test' (deprecated, use 'testRuntimeOnly ' instead).
    extends from: testCompile, runtime, testSupportRuntime
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar
         (*) build\dist\core-2.0.0-SNAPSHOT.jar

testRuntimeClasspath - Runtime classpath of source set 'test'.
    extends from: testRuntimeOnly, testRuntime, testImplementation
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar
         (*) build\dist\core-2.0.0-SNAPSHOT.jar

testRuntimeOnly - Runtime only dependencies for source set 'test'.
    extends from: runtimeOnly

testSupportAnnotationProcessor - Annotation processors and their dependencies for source set 'test support'.

testSupportCompile - Dependencies for source set 'test support' (deprecated, use 'testSupportImplementation ' instead).
    extends from: compile
    artifacts:
             build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
             build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testSupportCompileClasspath - Compile classpath for source set 'test support'.
    extends from: testSupportCompileOnly, testSupportImplementation
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testSupportCompileOnly - Compile only dependencies for source set 'test support'.

testSupportCompile_ConfigurationProxy_ForResolvingProjectDependencies - null
    extends from: testSupportCompile
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testSupportImplementation - Implementation only dependencies for source set 'test support'.
    extends from: testSupportCompile
    artifacts:
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testSupportRuntime - Runtime dependencies for source set 'test support' (deprecated, use 'testSupportRuntimeOnly ' instead).
    extends from: runtime, testSupportCompile
    artifacts:
         (*) build\dist\core-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testSupportRuntimeClasspath - Runtime classpath of source set 'test support'.
    extends from: testSupportRuntimeOnly, testSupportRuntime, testSupportImplementation
    artifacts:
         (*) build\dist\core-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT.jar
         (*) build\dist\core-testsupport-2.0.0-SNAPSHOT-sources.jar

testSupportRuntimeOnly - Runtime only dependencies for source set 'test support'.

(*) - inherited artifact

> Task :test:configurations
------------------------------------------------------------
Project ':test' - Common test framework that all projects should include as a testCompile dependency and all tests should extend BaseTest.
------------------------------------------------------------

annotationProcessor - Annotation processors and their dependencies for source set 'main'.

apiElements - API elements for main.
    extends from: runtime
    artifacts:
             build\dist\test-1.9.0-SNAPSHOT.jar

archives - Configuration for archive artifacts.
    artifacts:
             build\dist\test-1.9.0-SNAPSHOT.jar
             build\dist\test-1.9.0-SNAPSHOT-sources.jar

compile - Dependencies for source set 'main' (deprecated, use 'implementation ' instead).

compileClasspath - Compile classpath for source set 'main'.
    extends from: compileOnly, implementation

compileOnly - Compile only dependencies for source set 'main'.

default - Configuration for default artifacts.
    extends from: runtimeElements
    artifacts:
         (*) build\dist\test-1.9.0-SNAPSHOT.jar

implementation - Implementation only dependencies for source set 'main'.
    extends from: compile

intTestAnnotationProcessor - Annotation processors and their dependencies for source set 'int test'.

intTestCompile - Dependencies for source set 'int test' (deprecated, use 'intTestImplementation ' instead).
    extends from: testSupportCompile
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

intTestCompileClasspath - Compile classpath for source set 'int test'.
    extends from: intTestCompileOnly, intTestImplementation
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

intTestCompileOnly - Compile only dependencies for source set 'int test'.

intTestImplementation - Implementation only dependencies for source set 'int test'.
    extends from: intTestCompile
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

intTestRuntime - Runtime dependencies for source set 'int test' (deprecated, use 'intTestRuntimeOnly ' instead).
    extends from: testSupportRuntime, intTestCompile
    artifacts:
         (*) build\dist\test-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

intTestRuntimeClasspath - Runtime classpath of source set 'int test'.
    extends from: intTestRuntimeOnly, intTestRuntime, intTestImplementation
    artifacts:
         (*) build\dist\test-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

intTestRuntimeOnly - Runtime only dependencies for source set 'int test'.

jacocoAgent - The Jacoco agent to use to get coverage data.

jacocoAnt - The Jacoco ant tasks to use to get execute Gradle tasks.

licenses - null

runtime - Runtime dependencies for source set 'main' (deprecated, use 'runtimeOnly ' instead).
    extends from: compile
    artifacts:
             build\dist\test-1.9.0-SNAPSHOT.jar

runtimeClasspath - Runtime classpath of source set 'main'.
    extends from: runtimeOnly, runtime, implementation
    artifacts:
         (*) build\dist\test-1.9.0-SNAPSHOT.jar

runtimeElements - Elements of runtime for main.
    extends from: implementation, runtimeOnly, runtime
    artifacts:
             build\dist\test-1.9.0-SNAPSHOT.jar

runtimeOnly - Runtime only dependencies for source set 'main'.

testAnnotationProcessor - Annotation processors and their dependencies for source set 'test'.

testCompile - Dependencies for source set 'test' (deprecated, use 'testImplementation ' instead).
    extends from: compile, testSupportCompile
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testCompileClasspath - Compile classpath for source set 'test'.
    extends from: testCompileOnly, testImplementation
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testCompileOnly - Compile only dependencies for source set 'test'.

testImplementation - Implementation only dependencies for source set 'test'.
    extends from: testCompile, implementation
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testRuntime - Runtime dependencies for source set 'test' (deprecated, use 'testRuntimeOnly ' instead).
    extends from: testCompile, runtime, testSupportRuntime
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar
         (*) build\dist\test-1.9.0-SNAPSHOT.jar

testRuntimeClasspath - Runtime classpath of source set 'test'.
    extends from: testRuntimeOnly, testRuntime, testImplementation
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar
         (*) build\dist\test-1.9.0-SNAPSHOT.jar

testRuntimeOnly - Runtime only dependencies for source set 'test'.
    extends from: runtimeOnly

testSupportAnnotationProcessor - Annotation processors and their dependencies for source set 'test support'.

testSupportCompile - Dependencies for source set 'test support' (deprecated, use 'testSupportImplementation ' instead).
    extends from: compile
    artifacts:
             build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
             build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testSupportCompileClasspath - Compile classpath for source set 'test support'.
    extends from: testSupportCompileOnly, testSupportImplementation
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testSupportCompileOnly - Compile only dependencies for source set 'test support'.

testSupportImplementation - Implementation only dependencies for source set 'test support'.
    extends from: testSupportCompile
    artifacts:
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testSupportRuntime - Runtime dependencies for source set 'test support' (deprecated, use 'testSupportRuntimeOnly ' instead).
    extends from: runtime, testSupportCompile
    artifacts:
         (*) build\dist\test-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testSupportRuntimeClasspath - Runtime classpath of source set 'test support'.
    extends from: testSupportRuntimeOnly, testSupportRuntime, testSupportImplementation
    artifacts:
         (*) build\dist\test-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT.jar
         (*) build\dist\test-testsupport-1.9.0-SNAPSHOT-sources.jar

testSupportRuntimeOnly - Runtime only dependencies for source set 'test support'.

(*) - inherited artifact


BUILD SUCCESSFUL in 1s
2 actionable tasks: 2 executed
```

**Comment by eriwen** [comment]
Interesting idea for a Gradle plugin. 

Also, do the docs on [Inspecting Dependencies](https://docs.gradle.org/current/userguide/inspecting_dependencies.html) and [Build Scans](https://scans.gradle.com) help satisfy your needs here?

**Comment by nniesen** [comment]
"Inspecting Dependencies" is the input library dependencies. This reports on the configuration relationships and the artifact outputs. I haven't really messed with Build Scans so can't comment on those. The main use-cases for us are to see the setup and verify that the setup happened as expected.

**Comment by ov7a** [comment]
Fixed in Gradle 7.5
https://docs.gradle.org/7.5/release-notes.html
